### PR TITLE
KAFKA-4042: prevent DistributedHerder thread from dying from connector/task lifecycle exceptions

### PR DESCRIPTION
**Pull request by shikhar**
- `worker.startConnector()` and `worker.startTask()` can throw (e.g. `ClassNotFoundException`, `ConnectException`, or any other exception arising from the constructor of the connector or task class when we `newInstance()`), so add catch blocks around those calls from the `DistributedHerder` and handle by invoking `onFailure()` which updates the `StatusBackingStore`.
- `worker.stopConnector()` throws `ConnectException` if start failed causing the connector to not be registered with the worker, so guard with `worker.ownsConnector()`
- `worker.stopTasks()` and `worker.awaitStopTasks()` throw `ConnectException` if any of them failed to start and are hence not registered with the worker, so guard those calls by filtering the task IDs with `worker.ownsTask()`

### CODE REVIEW

**Comment by shikhar** [comment]
cc @ewencp @hachikuji for review

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java [hunk 1]
```diff
@@ -154,7 +154,7 @@ public void onShutdown(String connector) {
 
     @Override
     public void onFailure(String connector, Throwable cause) {
-        statusBackingStore.putSafe(new ConnectorStatus(connector, ConnectorStatus.State.FAILED,
+        statusBackingStore.put(new ConnectorStatus(connector, ConnectorStatus.State.FAILED,
```
Comment (shikhar) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: Previously `putSafe` was okay because connectors & tasks would always have some other state like `STARTED`, but it's possible now to transition directly to `FAILED`.

File: tests/kafkatest/services/connect.py [hunk 2]
```diff
@@ -361,22 +361,19 @@ def start(self):
 
 class MockSource(object):
 
-    def __init__(self, cc, topics, mode=None, delay_sec=10, name="mock-source"):
+    def __init__(self, cc, mode=None, delay_sec=10, name="mock-source"):
```
Comment (shikhar) [inline tests/kafkatest/services/connect.py:382]: unrelated minor change, topics didn't make sense for source connector

File: tests/kafkatest/tests/connect/connect_distributed_test.py [hunk 3]
```diff
@@ -155,7 +155,43 @@ def test_restart_failed_connector(self):
...
+        self.cc.set_configs(lambda node: self.render("connect-distributed.properties", node=node))
+        self.cc.start()
+
+        connector_name = 'bad-to-good-test'
+
+        connector = namedtuple('BadConnector', ['name', 'tasks'])(connector_name, 1)
+        config = {
+            'name': connector.name,
+            'tasks.max': connector.tasks,
+            'connector.class': 'java.util.HashMap'
```
Comment (shikhar) [inline tests/kafkatest/tests/connect/connect_distributed_test.py:175]: I have a not great reason for not using `NonExistentConnectorClass` here... if `ClassNotFoundException` is triggered in `Worker.startConnector()`, it falls back to reflection for traversing classpath to find a class matching that alias and that takes a while in system test VM's. Long enough that the 15s status check times out.

https://github.com/apache/kafka/blob/59cfa84801c67de9729385a8f9b536721e0c37b9/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java#L201

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java (hunk 1 above)
Comment (hachikuji) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: Interesting. I think the case I was concerned about when creating the poorly named `putSafe` methods was a defunct task which didn't shutdown quick enough and then ultimately transitioned itself to FAILED after the group had already rebalanced. Maybe the `cancelled` flag in `WorkerTask` provides enough protection from this case? Should we have something similar in `WorkerConnector`?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java [hunk 4]
```diff
@@ -573,9 +582,11 @@ public Void call() throws Exception {
                     return null;
                 }
 
-                if (worker.ownsConnector(connName)) {
+                if (assignment.connectors().contains(connName)) {
```
Comment (hachikuji) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:585]: I guess the reason for this check is that we catch exceptions starting the connector from the herder. I'm wondering whether you considered catching them in the worker instead?

Comment (shikhar) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:585]: The reason for handling exceptions in `DistributedHerder` rather than `Worker` - I figure the current behavior works well for `StandaloneHerder` which calls the same methods on `Worker`, since it should just crash if it runs into basic errors with the connector or its tasks.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaStatusBackingStore.java [hunk 5]
```diff
@@ -448,14 +448,14 @@ public boolean isDeleted() {
             return deleted;
         }
 
-        public boolean canWrite(T status) {
-            return value != null &&
```
Comment (shikhar) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaStatusBackingStore.java:452]: The `value != null` condition was contrary to the motivation stated in this class of allowing for a 'safe' write to happen if there is no previous value.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java (hunk 1 above)
Comment (shikhar) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: This change was misguided, I've reverted that and instead made a bugfix here https://github.com/apache/kafka/pull/1745/files#r75564769

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaStatusBackingStore.java (hunk 5 above)
Comment (shikhar) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaStatusBackingStore.java:452]: The motivation for this change is that connector/task status listeners' `onFailure()` callbacks invoke `StatusBackingListener.putSafe()`, but the status was not being reflected if there wasn't a previous value (e.g. the STARTED state, which is written via the `StatusBackingListener.put()` path) since `canWrite()` was returning false.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java [hunk 6]
```diff
@@ -301,10 +300,20 @@ private void processConnectorConfigUpdates(Set<String> connectorConfigUpdates) {
...
+            }
             // The update may be a deletion, so verify we actually need to restart the connector
-            if (remains)
-                startConnector(connectorName);
+            if (remains) {
+                try {
+                    startConnector(connectorName);
+                } catch (Throwable t) {
+                    log.warn("Failed to start connector {}", connectorName, t);
+                    onFailure(connectorName, t);
```
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:314]: Until now, it looks like we isolated the status changes to `Worker`, which makes them work automatically for both standalone and distributed mode. This patch seems to address distributed mode, but can't most of the same errors happen in `StandaloneHerder` too? I think this is related to @hachikuji's comment below too -- basically if we can push some of this handling into Worker then I think we'd keep the use of the `onFailure` better localized. It might actually also reduce some code redundancy between the two Herder implementations.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java (hunk 4 above)
Comment (ewencp) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:585]: Hmm, interesting. I think this kind of makes sense, but is also a bit weird if you run multiple connectors per process. I guess at that point you're just accepting that your tasks won't be isolated?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java [hunk 7]
```diff
@@ -1053,16 +1064,29 @@ public void onRevoked(String leader, Collection<String> connectors, Collection<C
                 // this worker. Instead, we can let them continue to run but buffer any update requests (which should be
                 // rare anyway). This would avoid a steady stream of start/stop, which probably also includes lots of
                 // unnecessary repeated connections to the source/sink system.
-                for (String connectorName : connectors)
-                    worker.stopConnector(connectorName);
+                for (String connectorName : connectors) {
+                    if (worker.ownsConnector(connectorName)) {
+                        worker.stopConnector(connectorName);
```
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:1069]: I think the other thing that's a bit confusing is how when you start a connector it needs try/catch, but stopping a connector doesn't (and handles its own failures). I get that this is because we have the extra `DistributedHerder.startConnector()` wrapper of `Worker.startConnector()`, but it feels asymmetric and confusing when reading this code.

**Comment by shikhar** [comment]
I'll be trying to move the error-handling into `WorkerConnector` and `WorkerTask`s, and make sure `Worker` does not throw aside from legit error conditions.

**Comment by shikhar** [comment]
Trying to contain exceptions in `WorkerConnector` or `WorkerTask` proved to be a very complicated refactoring, as a lot of early work needed to be delayed into their `initialize()` methods. I instead went with containing exceptions in the `Worker` (#1778) as suggested by @hachikuji and @ewencp.
